blockchain/blockchain.go

```diff
@@ -209,20 +270,87 @@ func getBlockByHash(txn db.Transaction, hash *felt.Felt) (block *core.Block, err
 	return
 }
 
-// SanityCheckNewHeight checks integrity of a block and resulting state update
-func (b *Blockchain) SanityCheckNewHeight(block *core.Block, stateUpdate *core.StateUpdate) error {
-	if !block.Hash.Equal(stateUpdate.BlockHash) {
-		return ErrIncompatibleBlockAndStateUpdate{errors.New("block hashes do not match")}
+type txAndReceiptDBKey struct {
+	Number uint64
+	Index  uint64
+}
+
+// storeTransactionAndReceipt stores the given transaction receipt in the database.
+// The db storage for transaction and receipts is maintained by three buckets as follows:
+//
+// [db.TransactionBlockNumbersAndIndicesByHash](TransactionHash) -> (BlockNumber, Index)
+// [db.TransactionsByBlockNumberAndIndex](BlockNumber, Index) -> Transaction
+// [db.ReceiptsByBlockNumberAndIndex](BlockNumber, Index) -> Receipt
+//
+// Note: we are using the same transaction hash bucket which keeps track of block number and
+// index for both transactions and receipts since transaction and its receipt share the same hash.
+// "[]" is the db prefix to represent a bucket
+// "()" are additional keys appended to the prefix or multiple values marshalled together
+// "->" represents a key value pair.
+func storeTransactionAndReceipt(txn db.Transaction, number, i uint64, t core.Transaction, r *core.TransactionReceipt) error {
+	bnIndexBytes, err := encoder.Marshal(txAndReceiptDBKey{number, i})
+	if err != nil {
+		return err
 	}
-	if !block.GlobalStateRoot.Equal(stateUpdate.NewRoot) {
-		return ErrIncompatibleBlockAndStateUpdate{
-			errors.New("block's GlobalStateRoot does not match state update's NewRoot"),
-		}
+
+	if err = txn.Set(db.TransactionBlockNumbersAndIndicesByHash.Key((r.TransactionHash).Marshal()),
+		bnIndexBytes); err != nil {
+		return err
 	}
 
-	if err := core.VerifyBlockHash(block, b.network); err != nil {
-		return ErrIncompatibleBlock{err}
+	if txnBytes, err := encoder.Marshal(t); err != nil {
+		return err
+	} else if err = txn.Set(db.TransactionsByBlockNumberAndIndex.Key(bnIndexBytes), txnBytes); err != nil {
+		return err
 	}
 
+	if rBytes, err := encoder.Marshal(r); err != nil {
+		return err
+	} else if err = txn.Set(db.ReceiptsByBlockNumberAndIndex.Key(bnIndexBytes), rBytes); err != nil {
+		return err
+	}
 	return nil
 }
+
+// getTransactionBlockNumberAndIndexByHash gets the block number and index for a given transaction hash
+func getTransactionBlockNumberAndIndexByHash(txn db.Transaction, hash *felt.Felt) (bnIndexBytes []byte, err error) {
+	err = txn.Get(db.TransactionBlockNumbersAndIndicesByHash.Key(hash.Marshal()), func(val []byte) error {
+		bnIndexBytes = val
```

Comment:
we have to copy val in to bnIndexBytes, simply assigning the slice is not valid.

Reply:
unmarshaling it in to a `txAndReceiptDBKey` works as well

Reply:
done

Reply:
> we have to copy val into bnIndexBytes, simply assigning the slice is not valid.

In a previous pull request, I raised this [point](https://github.com/NethermindEth/juno/pull/557#discussion_r1107078125) and you said because of callbacks val does leave scope of the callback so we didn't have to copy.

Also, do unmarshalling functions always create a copy?

Reply:
This also means that there is a bug in our get block functions because it follows the same previous pattern. I will look into it.

Reply:
I must have been unclear then.

If we want to use `val` outside of the callback, we must copy. Assume `val` is an array allocated on the stack. So, outside the callback it does not exist.

>  This also means that there is a bug in our get block functions because it follows the same previous pattern. I will look into it.

I see unmarshals only in our block getters, which one are you referring to?

Reply:
> Also, do unmarshalling functions always create a copy?

they allocate a new object and populate it from the given data, they don't use the underlying memory location that `val` points to